File: Week_07/547/friend_circles2.py
from typing import List


class Solution:
    def findCircleNum(self, M: List[List[int]]) -> int:
        n = len(M)
        res = 0
        parent = [-1] * n
        for i in range(n):
            for j in range(n):
                if i != j and M[i][j] == 1:
                    self.union(parent, i, j)
        for p in parent:
            if p == -1:
                res += 1

        return res

    def find(self, parent: List[int], i: int):
        if parent[i] == -1:
            return i
        return self.find(parent, parent[i])

    def union(self, parent: List[int], x: int, y: int):
        px = self.find(parent, x)
        py = self.find(parent, y)
        if px != py:
            parent[px] = py

File: Week_07/547/friend_circles2.go
package main

func findCircleNum2(M [][]int) int {
	n := len(M)
	res := 0

	parent := make([]int, n)
	for i := 0; i < n;i++ {
		parent[i] = -1
	}
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			if i != j && M[i][j] == 1 {
				union(parent, i, j)
			}
		}
	}
	for i := 0; i < n; i++ {
		if parent[i] == -1 {
			res++
		}
	}
	return res
}

func union(parent []int, x, y int) {
	px := find(parent, x)
	py := find(parent, y)
	if px != py {
		parent[px] = py
	}
}
func find(parent []int, i int) int {
	if parent[i] == -1 {
		return i
	}
	return find(parent, parent[i])
}
